Mathlib/FieldTheory/Differential/Liouville.lean

```diff
@@ -0,0 +1,207 @@
+      · rhs
+        intro x
+        rhs
+        apply logDeriv_algebraMap (u x)
+      · apply (deriv_algebraMap v)
+    simp_rw [IsScalarTower.algebraMap_apply F B K]
+    norm_cast
+
+
+/--
+Transfer an `IsLiouville` instance using an equivalence `K ≃ₐ[F] K'`.
+Requires an algebraic `K'` to show that the equivalence commutes with the derivative.
+-/
+lemma IsLiouville.equiv {K' : Type*} [Field K'] [Differential K'] [Algebra F K']
+    [DifferentialAlgebra F K'] [Algebra.IsAlgebraic F K']
+    [inst : IsLiouville F K] (e : K ≃ₐ[F] K') : IsLiouville F K' where
+  isLiouville (a : F) (ι : Type) [Fintype ι] (c : ι → F) (hc : ∀ x, (c x)′ = 0)
+      (u : ι → K') (v : K') (h : a = ∑ x, c x * logDeriv (u x) + v′) := by
+    apply inst.isLiouville a ι c hc (e.symm ∘ u) (e.symm v)
+    apply_fun e.symm at h
+    simpa [AlgEquiv.commutes, map_add, map_sum, map_mul, logDeriv, algEquiv_deriv'] using h
+
+/--
+A finite dimensional Galois extension of `F` is a Liouville extension.
+This is private because it's generalized by all finite dimensional extensions being Liouville.
+-/
+private local instance isLiouville_of_finiteDimensional_galois [FiniteDimensional F K]
+    [IsGalois F K] : IsLiouville F K where
+  isLiouville (a : F) (ι : Type) [Fintype ι] (c : ι → F) (hc : ∀ x, (c x)′ = 0)
+      (u : ι → K) (v : K) (h : a = ∑ x, c x * logDeriv (u x) + v′) := by
+    haveI : CharZero K := charZero_of_injective_algebraMap
+      (NoZeroSMulDivisors.algebraMap_injective F K)
+    -- We sum `e x` over all isomorphisms `e : K ≃ₐ[F] K`.
+    -- Because this is a Galois extension each of the relevant values will be in `F`.
+
+    -- We need to divide by `Fintype.card (K ≃ₐ[F] K)` to get the original answer.
+    let c' (i : ι) := (c i) / (Fintype.card (K ≃ₐ[F] K))
+    -- logDeriv turns sums to products, so the new `u` will be the product of the old `u` over all
+    -- isomorphisms
+    let u'' (i : ι) := ∏ x : (K ≃ₐ[F] K), x (u i)
+
+    -- Each of the values of u'' are fixed by all isomorphisms.
+    have : ∀ i, u'' i ∈ fixedField (⊤ : Subgroup (K ≃ₐ[F] K)) := by
+      rintro i ⟨e, _⟩
+      change e (u'' i) = u'' i
+      simp only [u'', map_prod]
+      apply Fintype.prod_equiv (Equiv.mulLeft e)
+      simp
+    have ffb : fixedField ⊤ = ⊥ := (IsGalois.tfae.out 0 1).mp (inferInstanceAs (IsGalois F K))
+    simp_rw [ffb, IntermediateField.mem_bot, Set.mem_range] at this
+
+    -- Therefore they are all in `F`. We use `choose` to get their values in `F`.
+    choose u' hu' using this
+
+    -- We do almost the same thing for `v''`, just with sum instead of product.
+    let v'' := (∑ x : (K ≃ₐ[F] K), x v) / (Fintype.card ((K ≃ₐ[F] K)))
+    have : v'' ∈ fixedField (⊤ : Subgroup (K ≃ₐ[F] K)) := by
+      rintro ⟨e, _⟩
+      change e v'' = v''
+      simp only [v'', map_div₀, map_sum, map_natCast]
+      congr 1
+      apply Fintype.sum_equiv (Equiv.mulLeft e)
+      simp
+    rw [ffb, IntermediateField.mem_bot] at this
+    obtain ⟨v', hv'⟩ := this
+
+
```

Comment:
I would prefer not using empty lines in proof blocks. Instead, write a comment explaining the new part of the proof coming up.
(In particular, please remove the empty lines above comments in the rest of the proof as well.)